src/Compiler/CMakeLists.txt

```diff
@@ -66,9 +66,36 @@ add_onnx_mlir_library(CompilerPasses
   )
 
 if (ONNX_MLIR_VENDOR)
-  set_source_files_properties(CompilerUtils.cpp
-    PROPERTIES COMPILE_DEFINITIONS "ONNX_MLIR_VENDOR=\"${ONNX_MLIR_VENDOR} \"")
+  list(APPEND DEFINITIONS "ONNX_MLIR_VENDOR=\"${ONNX_MLIR_VENDOR}\"")
 endif()
+if (DEFINED PRODUCT_VERSION_MAJOR)
+  if (NOT DEFINED PRODUCT_ID)
+    message(FATAL_ERROR "PRODUCT_ID must be defined with PRODUCT_VERSION_MAJOR")
+  endif()
+  if (PRODUCT_VERSION_MAJOR LESS_EQUAL 0)
+    message(FATAL_ERROR "PRODUCT_VERSION_MAJOR must be greater than zero")
+  endif()
+  list(APPEND DEFINITIONS "PRODUCT_VERSION_MAJOR=${PRODUCT_VERSION_MAJOR}")
+endif()
+if (DEFINED PRODUCT_VERSION_MINOR)
+  if (NOT DEFINED PRODUCT_VERSION_MAJOR)
+    message(FATAL_ERROR "PRODUCT_VERSION_MAJOR must be defined with PRODUCT_VERSION_MINOR")
+  endif()
+  list(APPEND DEFINITIONS "PRODUCT_VERSION_MINOR=${PRODUCT_VERSION_MINOR}")
+endif()
+if (DEFINED PRODUCT_VERSION_PATCH)
+  if (NOT DEFINED PRODUCT_VERSION_MINOR)
+    message(FATAL_ERROR "PRODUCT_VERSION_MINOR must be defined with PRODUCT_VERSION_PATCH")
+  endif()
+  list(APPEND DEFINITIONS "PRODUCT_VERSION_PATCH=${PRODUCT_VERSION_PATCH}")
+endif()
+if (PRODUCT_ID)
+  if (NOT DEFINED PRODUCT_VERSION_MAJOR)
+    message(FATAL_ERROR "PRODUCT_VERSION_MAJOR must be defined with PRODUCT_ID")
+  endif()
+  list(APPEND DEFINITIONS "PRODUCT_ID=\"${PRODUCT_ID}\"")
+endif()
+set_property(SOURCE CompilerUtils.cpp APPEND PROPERTY COMPILE_DEFINITIONS ${DEFINITIONS})
```

Comment:
Using 4 variables seems a bit cumbersome for this. Instead with regex parsing 1 variable such as `-DONNX_MLIR_VERSION=1.2.3-foo` would suffice:
```
# ONNX_MLIR_VERSION is specified/cached.
if (ONNX_MLIR_VERSION)
  string(REGEX MATCH "^([0-9]+)(\\.([0-9]+))?(\\.([0-9]+))?(\\-(.+))?$"
    VERSION_MATCH ${ONNX_MLIR_VERSION})

  # For a version string 1.2.3-foo
  #
  # CMAKE_MATCH_1 = 1
  # CMAKE_MATCH_2 = .2
  # CMAKE_MATCH_3 = 2
  # CMAKE_MATCH_4 = .3
  # CMAKE_MATCH_5 = 3
  # CMAKE_MATCH_6 = -foo
  # CMAKE_MATCH_7 = foo
  #
  # Major is mandatory, minor, patch, and suffix are optional.
  # When not specified, minor, patch, and suffix are empty
  # (or can default to some value)
  set(ONNX_MLIR_VERSION_MAJOR  ${CMAKE_MATCH_1})
  set(ONNX_MLIR_VERSION_MINOR  ${CMAKE_MATCH_3})
  set(ONNX_MLIR_VERSION_PATCH  ${CMAKE_MATCH_5})
  set(ONNX_MLIR_VERSION_SUFFIX ${CMAKE_MATCH_7})

  message(DEBUG "ONNX_MLIR_VERSION_MAJOR =${ONNX_MLIR_VERSION_MAJOR}")
  message(DEBUG "ONNX_MLIR_VERSION_MINOR =${ONNX_MLIR_VERSION_MINOR}")
  message(DEBUG "ONNX_MLIR_VERSION_PATCH =${ONNX_MLIR_VERSION_PATCH}")
  message(DEBUG "ONNX_MLIR_VERSION_SUFFIX=${ONNX_MLIR_VERSION_SUFFIX}")

  # Major must be specified if ONNX_MLIR_VERSION is specified/cached.
  if ("${ONNX_MLIR_VERSION_MAJOR}" STREQUAL "")
    message(STATUS "onnx-mlir version        : invalid version ${ONNX_MLIR_VERSION} ignored")
    unset(ONNX_MLIR_VERSION CACHE)
  else()
    message(STATUS "onnx-mlir version        : ${ONNX_MLIR_VERSION}")
    list(APPEND DEFINITIONS "ONNX_MLIR_VERSION_MAJOR=${ONNX_MLIR_VERSION_MAJOR}")
    if (NOT "${ONNX_MLIR_VERSION_MINOR}" STREQUAL "")
      list(APPEND DEFINITIONS "ONNX_MLIR_VERSION_MINOR=${ONNX_MLIR_VERSION_MINOR}")
    endif()
    if (NOT "${ONNX_MLIR_VERSION_PATCH}" STREQUAL "")
      list(APPEND DEFINITIONS "ONNX_MLIR_VERSION_PATCH=${ONNX_MLIR_VERSION_PATCH}")
    endif()
    if (NOT "${ONNX_MLIR_VERSION_SUFFIX}" STREQUAL "")
      list(APPEND DEFINITIONS "ONNX_MLIR_VERSION_SUFFIX=${ONNX_MLIR_VERSION_SUFFIX}")
    endif()
    set_property(SOURCE CompilerUtils.cpp APPEND PROPERTY COMPILE_DEFINITIONS ${DEFINITIONS})
  endif()
else()
  message(STATUS "onnx-mlir version        : not set")
endif()
```

Reply:
Good idea, I modified the code with this idea, but not exactly the same.